File: frontend/apps/web/src/apis/exhibition.ts
// apps/web/src/apis/exhibition.ts

import api from './index';

// =============================================================================
// [Type] 공통 / 서브 타입 (디자인, 영화, 키워드)
// =============================================================================

export interface ExhibitionDesign {
  font: string;
  colorScheme: string;
  cukeeStyle: string;
  frameStyle: string;
  background: string;
  backgroundImage: string;
}

export interface ExhibitionMovie {
  id: number;           // 전시회 내부 관리용 ID
  movieId: number;      // 실제 영화 ID
  title: string;
  posterUrl: string;
  displayOrder: number;
  isPinned: boolean;
  curatorComment: string;
}

export interface ExhibitionKeyword {
  keyword: string;
  weight: number;
}

// =============================================================================
// [Type] 전시회 (Exhibition) 관련 인터페이스
// =============================================================================

// 1. 전시회 목록 조회용 (간략 정보)
export interface Exhibition {
  id: number;
  ticketId: number;
  title: string;
  curator: string;      // 큐레이터 이름
  curatorMsg: string;   // 큐레이터 한마디
  likes: number;
  imageUrl: string;     // 대표 이미지

  // 선택적 속성 (UI 상황에 따라 필요)
  tags?: string[];
  color?: string;
  width?: number;
  height?: number;
}

// 2. 전시회 목록 응답
export interface ExhibitionListResponse {
  data: Exhibition[];
  total: number;
  page: number;
  limit: number;
}

// 3. 전시회 상세 조회용 (전체 정보)
export interface ExhibitionDetailResponse {
  id: number;
  userId: number;
  title: string;
  isPublic: boolean;

  // 상세 중첩 객체
  design: ExhibitionDesign;
  movies: ExhibitionMovie[];
  keywords: ExhibitionKeyword[];

  createdAt: string;
  updatedAt: string;
}

// 전시회 꾸미기 
export type FrameStyle = 'none' | 'frame'
export type bgStyle = 'white' | 'pink' | 'orange' | 'pattern' | 'image'
export type CukeeStyle = 'line' | 'none' | 'vintage'

export interface ExhibitionStyle {
  frame: FrameStyle
  background: bgStyle
  cukee: CukeeStyle
}

// =============================================================================
// [Type] 티켓 (Ticket) 관련 인터페이스
// =============================================================================

export interface Ticket {
  id: number;
  title: string;
  curatorName: string;
  tags: string[];
  ticketImageUrl: string;
  characterImageUrl: string | null;
  width: number;
  height: number;
  ticketCode: string; // 예: 'shortform_mz'
  curatorMessage: string | null;
  color: string | null;
  description: string | null;
  likeCount: number;
  isLiked: boolean;
}

export interface TicketListResponse {
  data: Ticket[];
  total: number;
}

export type TicketDetailResponse = Ticket;

// =============================================================================
// [API Functions] 전시회
// =============================================================================

/**
 * 전시회 목록 조회
 * GET /exhibitions
 */
export const getExhibitions = async (page = 1, limit = 20): Promise<ExhibitionListResponse> => {
  const response = await api.get<ExhibitionListResponse>('/exhibitions', {
    params: { page, limit },
  });
  return response.data;
};

/**
 * 전시회 상세 조회
 * GET /exhibitions/{id}
 */
export const getExhibitionDetail = async (id: number): Promise<ExhibitionDetailResponse> => {
  const response = await api.get<ExhibitionDetailResponse>(`/exhibitions/${id}`);
  return response.data;
};

/**
 * 전시회 생성
 * POST /exhibitions_new
 */
export interface CreateExhibitionRequest {
  title: string;
  isPublic?: boolean;
  movies?: Array<{
    movieId: number;
    displayOrder: number;
    isPinned?: boolean;
  }>;
}

export interface CreateExhibitionResponse {
  id: number;
  userId: number;
  title: string;
  isPublic: boolean;
  createdAt: string;
  updatedAt: string;
}

export const createExhibition = async (
  data: CreateExhibitionRequest
): Promise<CreateExhibitionResponse> => {
  const response = await api.post<CreateExhibitionResponse>('/exhibitions', data);
  return response.data;
};

/**
 * 내 전시회 목록 조회 (user_id 필터링)
 * GET /exhibitions?user_id={current_user}
 */
export const getMyExhibitions = async (page = 1, limit = 20): Promise<ExhibitionListResponse> => {
  // user_id는 백엔드에서 인증 토큰으로 자동 판별하므로 따로 전달 안 함
  const response = await api.get<ExhibitionListResponse>('/exhibitions', {
    params: { page, limit },
  });
  return response.data;
};

// 전시회 상세 조회
export const getExhibitionById = async (exhibitionId: number) => {
  const response = await api.get(`/exhibitions/${exhibitionId}`);
  return response.data;
};

/**
 * 전시회 수정
 * PUT /exhibitions/{id}
 */
export interface UpdateExhibitionRequest {
  title?: string;
  isPublic?: boolean;
}

export const updateExhibition = async (
  exhibitionId: number,
  data: UpdateExhibitionRequest
): Promise<CreateExhibitionResponse> => {
  const response = await api.put<CreateExhibitionResponse>(`/exhibitions/${exhibitionId}`, data);
  return response.data;
};

/**
 * 전시회 삭제
 * DELETE /exhibitions/{id}
 */
export const deleteExhibition = async (exhibitionId: number): Promise<void> => {
  await api.delete(`/exhibitions/${exhibitionId}`);
};

// =============================================================================
// [API Functions] 티켓
// =============================================================================

/**
 * 티켓 전체 목록 조회
 * GET /tickets
 */
export const fetchTickets = async (): Promise<TicketListResponse> => {
  const response = await api.get<TicketListResponse>('/tickets');
  return response.data;
};

/**
 * 티켓 상세 조회
 * GET /tickets/{ticket_code}
 */
export const fetchTicketDetail = async (ticketCode: string): Promise<TicketDetailResponse> => {
  const response = await api.get<TicketDetailResponse>(`/tickets/${ticketCode}`);
  return response.data;
};

/**
 * 티켓 좋아요 토글
 * POST /tickets/{ticket_id}/like
 */
export interface ToggleLikeResponse {
  success: boolean;
  isLiked: boolean;
  likeCount: number;
}

export const toggleTicketLike = async (ticketId: number): Promise<ToggleLikeResponse> => {
  const response = await api.post<ToggleLikeResponse>(`/tickets/${ticketId}/like`);
  return response.data;
};

File: backend/app/schemas/exhibition.py
"""
전시회 관련 Pydantic 스키마
"""
from pydantic import BaseModel, ConfigDict, Field
from typing import List, Optional
from datetime import datetime


class ExhibitionListItem(BaseModel):
    """전시회 목록 아이템 (홈 화면용)"""
    id: int
    ticketId: int
    title: str
    curator: str
    curatorMsg: str
    likes: int
    imageUrl: str
    tags: List[str]
    color: str
    width: int
    height: int

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "ticketId": 1,
                "title": "숏폼 러버 MZ",
                "curator": "트렌디 큐레이터 MZ",
                "curatorMsg": "빠르게 즐기는 MZ세대 맞춤 영화",
                "likes": 150,
                "imageUrl": "https://example.com/ticket1.jpg",
                "tags": ["트렌디", "숏폼", "MZ"],
                "color": "#FF6B6B",
                "width": 300,
                "height": 450
            }
        }
    )


class ExhibitionListResponse(BaseModel):
    """전시회 목록 응답"""
    data: List[ExhibitionListItem]
    total: int
    page: int
    limit: int

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "data": [],
                "total": 11,
                "page": 1,
                "limit": 20
            }
        }
    )


class ExhibitionDesign(BaseModel):
    """전시회 디자인"""
    font: str
    colorScheme: str
    layoutType: str
    frameStyle: str
    background: str
    backgroundImage: str

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionMovie(BaseModel):
    """전시회 영화"""
    id: int
    movieId: int
    title: str
    posterUrl: str
    displayOrder: int
    isPinned: bool
    curatorComment: str

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionKeyword(BaseModel):
    """전시회 키워드"""
    keyword: str
    weight: float

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionDetailResponse(BaseModel):
    """전시회 상세 응답"""
    id: int
    userId: int
    title: str
    isPublic: bool
    design: ExhibitionDesign
    movies: List[ExhibitionMovie]
    keywords: List[ExhibitionKeyword]
    createdAt: datetime
    updatedAt: datetime

    model_config = ConfigDict(populate_by_name=True)


# ===== CRUD 스키마 =====

class ExhibitionDesignCreate(BaseModel):
    """전시회 디자인 생성 스키마"""
    font: Optional[str] = Field(None, max_length=50)
    colorScheme: Optional[str] = Field(None, max_length=50, alias="color_scheme")
    cukeeStyle: Optional[str] = Field(None, max_length=50, alias="cukee_style")
    frameStyle: Optional[str] = Field(None, max_length=20, alias="frame_style")
    background: Optional[str] = Field(None, max_length=100)
    backgroundImage: Optional[str] = Field(None, alias="background_image")

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionMovieCreate(BaseModel):
    """전시회 영화 생성 스키마"""
    movieId: int = Field(..., alias="movie_id")
    displayOrder: int = Field(0, alias="display_order")
    curatorComment: Optional[str] = Field(None, alias="curator_comment")
    isPinned: bool = Field(False, alias="is_pinned")

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionKeywordCreate(BaseModel):
    """전시회 키워드 생성 스키마"""
    keyword: str = Field(..., max_length=100)
    weight: float = Field(1.0)

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionCreate(BaseModel):
    """전시회 생성 요청"""
    title: str = Field(..., min_length=1, max_length=200, description="전시회 제목")
    isPublic: bool = Field(default=True, description="공개 여부")
    ticketId: Optional[int] = Field(default=None, description="티켓 ID (큐레이터 정보)")
    movies: List[ExhibitionMovieCreate] = Field(default_factory=list, description="영화 목록")
    keywords: List[ExhibitionKeywordCreate] = Field(default_factory=list, description="키워드 목록")
    design: Optional[ExhibitionDesignCreate] = Field(default=None, description="디자인 설정")

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={
            "example": {
                "title": "나만의 영화 전시회",
                "isPublic": True,
                "ticketId": 1,
                "movies": [
                    {"movieId": 123, "displayOrder": 0, "curatorComment": "추천 영화", "isPinned": True}
                ],
                "keywords": [
                    {"keyword": "로맨스", "weight": 0.8}
                ]
            }
        }
    )


class ExhibitionUpdate(BaseModel):
    """전시회 수정 요청"""
    title: Optional[str] = Field(None, max_length=200)
    isPublic: Optional[bool] = None
    design: Optional[ExhibitionDesignCreate] = None
    movies: Optional[List[ExhibitionMovieCreate]] = None
    keywords: Optional[List[ExhibitionKeywordCreate]] = None

    model_config = ConfigDict(populate_by_name=True)


class ExhibitionResponse(BaseModel):
    """전시회 기본 응답"""
    id: int
    userId: int
    title: str
    isPublic: bool
    createdAt: datetime
    updatedAt: datetime

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)
